### PR TITLE
Follow absolute and relative redirects when option is set to :follow_…

### PR DESCRIPTION
**Pull request by thilko**
Hi,

I am using httpotion for my [pinboard client](http://www.github.com/thilko/expinboard) and I need the follow redirect feature. I tried to implement it, but I must say that I am very new to the elixir world. What do you think about it?

### CODE REVIEW

File: lib/httpotion.ex [hunk 1]
```diff
@@ -120,18 +135,32 @@ defmodule HTTPotion.Base do
...
           :ibrowse.send_req(args[:url], args[:headers], args[:method], args[:body], args[:ib_options], args[:timeout])
-        end |> handle_response
+        end
+
+        if response_ok(response) && is_redirect(response) && options[:follow_redirects] do
+          location = process_response_location(response)
+          next_url = if String.starts_with?(location, "http") do
+            location
+          else
+            Regex.named_captures(~r/(?<url>http:\/\/.*?)\//, url)["url"] <> location
```
Comment (myfreeweb) [inline lib/httpotion.ex]: This will miss `https` URLs

**Comment by myfreeweb** [comment]
Looks nice, thank you for doing this! :-)

How does it interact with async requests?

**Comment by thilko** [comment]
> 
> Looks nice, thank you for doing this! :-)
> 
> How does it interact with async requests?

Ah, good point, I missed this as well as https urls. I´ll have a look at it. Thx!

**Comment by thilko** [comment]
I am afraid I don´t know how to follow the redirects the asynchronous way. I had a look at the `transformer` method, and in case of an redirect http status code I called the `request` method with the redirect url again. But the written tests are still failing, maybe because the first request is returned. Would it be an option to send the redirect request synchronously?

**Comment by myfreeweb** [comment]
Hmm… I don't see the code, but I think I know the problem.

When the `transformer` process receives `:ibrowse_async_headers`, you detect a redirect there.

Do not recursively call `transformer(target)` anymore, i.e. quit the process, or it will forward the chunk/end for the redirect response! (And do not send the `AsyncHeaders` for the redirect either, obviously.)

**Comment by thilko** [comment]
Got the test running now with follow redirects for async requests. Could you have a look on it? I am not sure about idiomatic elixir...

**Comment by myfreeweb** [comment]
Looks like you're still sending the chunks and end from the original 3xx response!

After you detect 3xx: you should call `request` *with* the `options` and NOT recursively call transformer(target) anymore. So, a new transformer process should be spawned, and the old one should stop.

**Comment by thilko** [comment]
I understand. But how do identify an 3xx response chunk? I don´t have the response code there, do I?

**Comment by myfreeweb** [comment]
You're already doing it: `if(process_status_code(status_code) in [302, 304])`

Inside of that `if` you should spawn a new transformer and not continue the old one.

**Comment by thilko** [comment]
Changed it. Now I got the responses with different id. Is that OK?

**Comment by myfreeweb** [comment]
Looks great now :-) Thank you very much.

**Pull request merged**